kernel/examples/read-table-multi-threaded/src/main.rs

```diff
@@ -0,0 +1,256 @@
+                .iter()
+                .map(|col| {
+                    table_schema
+                        .field(col)
+                        .cloned()
+                        .ok_or(deltakernel::Error::Generic(format!(
+                            "Table has no such column: {col}"
+                        )))
+                })
+                .try_collect();
+            selected_fields.map(|selected_fields| Arc::new(Schema::new(selected_fields)))
+        })
+        .transpose()?;
+
+    // build a scan with the specified schema
+    let scan = ScanBuilder::new(snapshot)
+        .with_schema_opt(read_schema_opt)
+        .build();
+
+    // this gives us an iterator of `ScanFile`s, which are just paths and metadata for the files
+    // that make up the table
+    let scan_files = scan.scan_files(engine_interface.as_ref())?;
+
+    // create the channels we'll use. record_batch_[t/r]x are used for the threads to send back the
+    // processed RecordBatches to themain thread
+    let (record_batch_tx, record_batch_rx) = mpsc::channel();
+    // scan_file_[t/r]x are used to send each scan file from the iterator out to the waiting threads
+    let (mut scan_file_tx, scan_file_rx) = spmc::channel();
+
+    // fire up each thread. we don't need the handles as we rely on the channels to indicate when
+    // things are done
+    let _handles: Vec<_> = (0..cli.thread_count)
+        .map(|_| {
+            // items that we need to send to the other thread
+            let interface = cli.interface.clone();
+            let scan_state = scan.get_scan_state().into_owned();
+            let url = url.clone();
+            let rb_tx = record_batch_tx.clone();
+            let scan_file_rx = scan_file_rx.clone();
+            thread::spawn(move || {
+                do_work(interface, scan_state, url, rb_tx, scan_file_rx);
+            })
+        })
+        .collect();
+
+    // have handed out all copies needed, drop so record_batch_rx will exit when the last thread is
+    // done sending
+    drop(record_batch_tx);
+
+    // send out each scan file
+    for scan_file in scan_files {
+        let scan_file = scan_file?;
+        scan_file_tx.send(scan_file).unwrap();
+    }
+
+    // have sent all scan files, drop this so threads will exit when there's no more work
+    drop(scan_file_tx);
+
+    // simply gather up each batch and print them
+    let mut batches = vec![];
+    for received in record_batch_rx {
+        batches.push(received);
+    }
+    print_batches(&batches)?;
+    Ok(())
+}
+
+// this is the work each thread does
+fn do_work(
+    interface: Interface, // what type of engine_interface was requested
+    scan_state: ScanState<'_>,
+    url: Url,
+    record_batch_tx: Sender<RecordBatch>,
+    scan_file_rx: spmc::Receiver<ScanFile>,
+) {
+    // todo: return a result and don't unwrap everywhere
+    // each thread needs its own copy since engine_interface isn't Clone, Send, or Sync
```

Comment:
We're not using an Arc here? 

The issue is that `EngineInterface` looks like:
```rust
pub trait EngineInterface {
  ...
}
``` 

and not
```rust
pub trait EngineInterface: Send + Sync {
  ...
}
```

So we can't take the interface created in `main()` and send it over here.

Interestingly we _do_ have:
```rust
pub trait ParquetHandler: Send + Sync {
   ...
}
```

and also `FileSystemClient` is `Send + Sync`, but `JsonHandler`, and `ExpressionHandler` are _not_.

Rather than make that change here I figured I'd just do this for now, but we should probably consider just making it all `Send + Sync`

Reply:
> We're not using an Arc here?

Not right here... but AFAIK every instance of `EngineInterface` we ever create is wrapped in `Arc`?

(but fair enough to revisit this situation in a separate PR)

Reply:
filed https://github.com/delta-incubator/delta-kernel-rs/issues/171

Reply:
I actually just did this in this PR btw, because it was trivial